gh-133157: remove use of `_Py_NO_SANITIZE_UNDEFINED` in faulthandler

picnixz opened the pull request:
- remove `_read_null()` as it's a tricky one (and the compiler didn't seem to complain)
- instead of using 1/0 arithmetic, we explicitly raise SIGFPE

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNNN: Summary of the changes made
```

Where: gh-NNNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNNNN)
```

Where: [X.Y] is the branch name, for example: [3.13].

GH-NNNNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-133157 -->
* Issue: gh-133157
<!-- /gh-issue-number -->

picnixz committed:
fix UBSan failures in `faulthandler.c`

- reading from NULL is no more an undefined behavior for C11
- instead of using 1/0 arithmetic, we explicitly raise SIGFPE

vstinner commented on Modules/faulthandler.c:
```diff
@@ -1158,23 +1157,13 @@ faulthandler_fatal_error_c_thread(PyObject *self, PyObject *args)
-
-    /* If the division by zero didn't raise a SIGFPE (e.g. on PowerPC),
-       raise it manually. */
+    /* raise SIGFPE manually to prevent crafted undefined behaviors */
```
I don't think that this comment makes sense for the new code. It can be removed.

picnixz replied:
```suggestion
```

vstinner commented on Modules/faulthandler.c:
```diff
@@ -1069,8 +1069,8 @@ faulthandler_suppress_crash_report(void)
-static PyObject* _Py_NO_SANITIZE_UNDEFINED
-faulthandler_read_null(PyObject *self, PyObject *args)
+static PyObject*
+faulthandler_read_null(PyObject *self, PyObject *Py_UNUSED(args))
```
Maybe read_null() should just be removed. Its test is quite complicated, I don't think that it's worth it:

```py
    @unittest.skipIf(sys.platform.startswith('aix'),
                     "the first page of memory is a mapped read-only on AIX")
    def test_read_null(self):
        if not MS_WINDOWS:
            self.check_fatal_error("""
                import faulthandler
                faulthandler.enable()
                faulthandler._read_null()
                """,
                3,
                # Issue #12700: Read NULL raises SIGILL on Mac OS X Lion
                '(?:Segmentation fault'
                    '|Bus error'
                    '|Illegal instruction)')
        else:
            self.check_windows_exception("""
                import faulthandler
                faulthandler.enable()
                faulthandler._read_null()
                """,
                3,
                'access violation')
```

faulthandler only cares about signals.

picnixz replied:
Since you're the faulthandler's maintainer, up to you! AFAICT, it's purpose is to check that faulthandler will catch errors when reading from NULL right? but this situation might occur though?

vstinner replied:
faulthandler registers signal handlers. It should not check for low-level C internals about undefined behaviors like reading from NULL.

picnixz replied:
Ok, then we can remove it I think. I'll do it tomorrow as I'm not on my dev session.

picnixz committed:
Update Modules/faulthandler.c

picnixz committed:
remove `faulthandler._read_null()`

Reading from NULL is an undefined behavior and `faulthandler`
should not check for low-level C about undefined behaviors.

vstinner reviewed (approved):
LGTM
